src/proxy/server.rs

```diff
@@ -190,164 +188,155 @@ impl fmt::Display for NoOriginalDst {
         let log = logging::Server::proxy(proxy_name, listen_addr);
-        Server {
-            drain_signal,
+        Self {
             http: hyper::server::conn::Http::new(),
+            h2_settings,
             listen_addr,
             accept,
             connect,
-            route,
+            make_http,
             log,
         }
     }
+}
 
-    pub fn log(&self) -> &logging::Server {
-        &self.log
-    }
-
+impl<A, T, C, H, B> ServeConnection<Connection> for Server<A, T, C, H, B>
+where
+    A: Accept<Connection> + Send + 'static,
+    A::Io: fmt::Debug + Send + Peek + 'static,
+    T: From<SocketAddr> + Send + 'static,
+    C: Service<T> + Clone + Send + 'static,
+    C::Response: AsyncRead + AsyncWrite + fmt::Debug + Send + 'static,
+    C::Future: Send + 'static,
+    C::Error: Into<Error>,
+    H: MakeService<
+            Source,
+            http::Request<HttpBody>,
+            Response = http::Response<B>,
+            MakeError = Never,
+        > + Clone
+        + Send
+        + 'static,
+    H::Error: Into<Error> + Send + 'static,
+    H::Service: Send + 'static,
+    H::Future: Send + 'static,
+    <H::Service as Service<http::Request<HttpBody>>>::Future: Send + 'static,
+    B: hyper::body::Payload + Default + Send + 'static,
+{
     /// Handle a new connection.
     ///
     /// This will peek on the connection for the first bytes to determine
     /// what protocol the connection is speaking. From there, the connection
     /// will be mapped into respective services, and spawned into an
     /// executor.
-    pub fn serve(
-        &self,
+    fn serve_connection(
+        &mut self,
         connection: Connection,
-        remote_addr: SocketAddr,
-        h2_settings: H2Settings,
-    ) -> impl Future<Item = (), Error = ()> {
-        let orig_dst = connection.original_dst_addr();
-        let disable_protocol_detection = !connection.should_detect_protocol();
-
-        let log = self.log.clone().with_remote(remote_addr);
-
+        drain: drain::Watch,
+    ) -> Box<dyn Future<Item = (), Error = ()> + Send + 'static> {
         let source = Source {
-            remote: remote_addr,
+            remote: connection.remote_addr(),
             local: connection.local_addr().unwrap_or(self.listen_addr),
-            orig_dst,
+            orig_dst: connection.original_dst_addr(),
             tls_peer: connection.peer_identity(),
             _p: (),
         };
 
+        let should_detect = connection.should_detect_protocol();
         let io = self.accept.accept(&source, connection);
 
-        let connect = self.connect.clone();
-
-        if disable_protocol_detection {
-            trace!("protocol detection disabled for {:?}", orig_dst);
-            let fwd = tcp::forward(io, connect, source);
-            let fut = self.drain_signal.clone().watch(fwd, |_| {});
-            return log.future(Either::B(fut));
-        }
+        let accept_fut = if should_detect {
+            Either::A(
```

Comment:
Nit/TIOLI: IMO, this might read better as 
```rust
let accept_fut = if should_detect {
    let f = io.peek()
       .map_err(|e| debug!("peek error: {}", e))
       .map(|io| {
            ...
        });
    Either::A(f)
} else { 
   ... 
}
```
this way, I have to figure out slightly fewer grouping symbols.

Not a big deal though.

Reply:
I don't disagree. I'm going to have be back in this file soon and will make polish changes then